### DISCUSSION TITLE
Getting previous build pull request

### DISCUSSION THREAD

**Discussion by AshishNigamNoida** [rfennell/AzurePipelines#1083] [General]
Hi , 

When i generate a new build to generate release notes it picked all the previous pull request. Ideally it should get only the pull requests associated with that particular build not all?
Could please let me know what setting is required for that?

Thanks!

Regards,
Ashish

**Comment by rfennell** [comment]
Commits and WI are found based on any created/associated since the last successful running of the build definition. It is from these that a list of PRs are found (by looking for specific fields)

You can see the logic being used if you set a variable `system.debug=true` or check the run with diagnostics checkbox on the pipeline run.  This will show you the build that is being compare against and what associations it found

You can upload the log here and I will have a look if you can't spot anything

**Reply by AshishNigamNoida** [reply]
[release.txt](https://github.com/rfennell/AzurePipelines/files/7179172/release.txt)

Please find the log

**Comment by rfennell** [comment]
That is the basic logging, it does not show all the debug details (you need the `system.debug=true` variable for that). 

Anyway, your current build is associated with 5 commits. This information is found using the Azure DevOps REST API call 

`buildApi.getBuildChanges(teamProject, buildId, "", 5000);` 

so this is nothing I have any control over it is the same call as used by the Azure DevOps UI

> 2021-09-16T14:35:28.0907373Z Getting the current build details
> 2021-09-16T14:35:28.1586289Z Getting items associated with only the current build
> 2021-09-16T14:35:28.3261983Z Get the artifacts consumed by the build
> 2021-09-16T14:35:28.3661327Z Cannot get extra commit ir work item details of the Repository artifact DevOps null
> 2021-09-16T14:35:28.3663444Z Get the file details associated with the commits
> 2021-09-16T14:35:28.3673556Z Enriched change c7397f379732b1d33899a337ca4442b87912e872 of type TfsGit
> 2021-09-16T14:35:28.4288990Z Enriched with details of 2 files
> 2021-09-16T14:35:28.4290848Z Enriched change d0476722a459902a7ff9e87ff9566ea8b353d1a3 of type TfsGit
> 2021-09-16T14:35:28.4626978Z Enriched with details of 2 files
> 2021-09-16T14:35:28.4628431Z Enriched change b4675cf0835d6efc733f17d4494a58718c4fef62 of type TfsGit
> 2021-09-16T14:35:28.4967934Z Enriched with details of 2 files
> 2021-09-16T14:35:28.4970027Z Enriched change ae8acb970526662f912bc929b69f6f93d339625d of type TfsGit
> 2021-09-16T14:35:28.5888181Z Enriched with details of 5 files
> 2021-09-16T14:35:28.5889537Z Enriched change a88c68ccc51a96dae7a5448419c90572a92d1268 of type TfsGit
> 2021-09-16T14:35:28.6527081Z Enriched with details of 7 files

Each of these commits is associated with PR hence what is returned.

> 2021-09-16T14:35:29.4079137Z Checking for PRs associated with the commit c7397f379732b1d33899a337ca4442b87912e872
> 2021-09-16T14:35:29.4080638Z - PR 7228 matches the commit c7397f379732b1d33899a337ca4442b87912e872
> 2021-09-16T14:35:29.4081859Z Checking for PRs associated with the commit d0476722a459902a7ff9e87ff9566ea8b353d1a3
> 2021-09-16T14:35:29.4082895Z - PR 7197 matches the commit d0476722a459902a7ff9e87ff9566ea8b353d1a3
> 2021-09-16T14:35:29.4083805Z Checking for PRs associated with the commit b4675cf0835d6efc733f17d4494a58718c4fef62
> 2021-09-16T14:35:29.4084991Z - PR 7106 matches the commit b4675cf0835d6efc733f17d4494a58718c4fef62
> 2021-09-16T14:35:29.4088893Z Checking for PRs associated with the commit ae8acb970526662f912bc929b69f6f93d339625d
> 2021-09-16T14:35:29.4090075Z - PR 7128 matches the commit ae8acb970526662f912bc929b69f6f93d339625d
> 2021-09-16T14:35:29.4090965Z Checking for PRs associated with the commit a88c68ccc51a96dae7a5448419c90572a92d1268
> 2021-09-16T14:35:29.4091885Z - PR 6895 matches the commit a88c68ccc51a96dae7a5448419c90572a92d1268

One thought, is this the first succesful running of this pipeline. If it is then you expect the release notes to call all past associated items in the repo

**Reply by AshishNigamNoida** [reply]
Suppose I have these builds and their Associated pull requests in my build stage 

Buid No #123

PullRequest #3
PullRequest #4

BuildNo #122

PullRequest #1
PullRequest #2

Issue is -In Release Note for Buid No#123 I am getting 4 assoicated pull request like below.Ideally It should 2

PullRequest #1
PullRequest #2
PullRequest #3
PullRequest #4

**Comment by rfennell** [comment]
As I said I cannot offer any other suggestions as I get the associaed commits using the SDK API call which is in effect just the REST endpoint 

> https://docs.microsoft.com/en-us/rest/api/azure/devops/build/builds/get-build-changes?view=azure-devops-rest-6.0

All I do is pass in the build ID and the list of associated commits is returned. These will match commits listed as associations with the Azure DevOps UI e.g this build is associated with 10 commits

![image](https://user-images.githubusercontent.com/324718/133648879-88898ec0-7057-443a-b41d-27a32f36353c.png)

Obviously there could be an issue with my PR association logic, but as your log shows it find 5 commits, not 2, I don't think the issue is with the PRs but the the Commits.

Maybe have a play with the API call in a browser to confirm what is being returned?
